Array Field – add index to edit aria label; allow removal of first item when not only item

gmrabian opened the pull request:
## Summary

- Reuse `uiItemName` variable and make lowercase where presented after capitalized text
- Rename per-array-item `itemName` variable to `ariaItemName` since that is its primary use
- Allow removal of the first array item if there are more than one items present

See comments for implementation details

## Related issue(s)

- https://app.zenhub.com/workspaces/vagov-product-team-forms-634853151f5c6000165942bc/issues/gh/department-of-veterans-affairs/va.gov-team/60371

## Testing done

- Ran unit tests
- Tested locally with screen reader

## Screenshots

### Before
aria label for first edit button (no distinction)
<img width="767" alt="Screenshot 2023-06-16 at 1 19 37 PM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/57802560/66f58ecf-5490-4051-a1db-18d7322a1cae">

aria label for fourth edit button (no distinction)
<img width="732" alt="Screenshot 2023-06-16 at 1 19 57 PM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/57802560/cb478829-8a69-43b4-a9ca-bea6c4eb2b53">

Item name examples (uppercase `Treatment record` in middle of sentence):
<img width="237" alt="Screenshot 2023-06-16 at 1 20 11 PM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/57802560/8b97d35a-3e99-44da-ae8e-e28f249a7fa4">
<img width="600" alt="Screenshot 2023-06-16 at 1 20 16 PM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/57802560/6d42eca1-d70f-49bf-a7f8-848d7bee6a56">
<img width="287" alt="Screenshot 2023-06-16 at 1 20 20 PM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/57802560/0c5ae3c8-4bce-41fa-b46c-ccad876f6b54">

### After
aria label for first edit button
<img width="838" alt="Screenshot 2023-06-16 at 2 53 37 PM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/57802560/813a081c-0c5d-4b29-a96f-5fb999368274">


aria label for fourth edit button
<img width="913" alt="Screenshot 2023-06-16 at 2 54 15 PM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/57802560/0b82603d-3665-437f-996e-340ea9bb4d6d">


Lowercase item name examples (`treatment record` vs `Treatment record`):
<img width="215" alt="Screenshot 2023-06-16 at 1 17 06 PM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/57802560/6fbe084c-c7bb-48be-abd2-8f55d9c786c7">
<img width="605" alt="Screenshot 2023-06-16 at 1 17 14 PM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/57802560/69cd7358-a67d-451c-9a4b-ad1a457d7d20">
<img width="270" alt="Screenshot 2023-06-16 at 1 17 22 PM" src="https://github.com/department-of-veterans-affairs/vets-website/assets/57802560/78e880b8-a2c9-410a-8039-d518307fe683">



## What areas of the site does it impact?
Enhances ArrayField experience with better aria labels and ability to remove first item when more items present

gmrabian committed:
add index to edit aria label; allow removal of first item when not only item

gmrabian committed:
remove underscore

gmrabian committed:
add indexes to edit and remove; make itemName lowercase when not leading text

gmrabian committed:
remove form dev tools

gmrabian committed:
Merge branch 'main' into simple-forms/update-array-field

gmrabian commented:
Unit test failure is happening in main too and is unrelated to these changes

gmrabian committed:
refactor for simplicity

gmrabian commented on src/platform/forms-system/src/js/fields/ArrayField.jsx:
```diff
@@ -509,7 +509,7 @@ export default class ArrayField extends React.Component {
             onClick={this.handleAdd}
           >
-            Add another {uiOptions.itemName}
+            Add another {(uiOptions.itemName || 'item').toLowerCase()}
```
this references `uiOptions.itemName` rather than just `itemName` because it's outside of the mapper where we have access to `itemName`

gmrabian commented on src/platform/forms-system/src/js/fields/ArrayField.jsx:
```diff
@@ -451,10 +453,8 @@ export default class ArrayField extends React.Component {
-                        'item'}?`}
-                      primaryButtonText={`Yes, remove this ${uiOptions.itemName ||
-                        'item'}`}
+                      modalTitle={`Are you sure you want to remove this ${itemName}?`}
+                      primaryButtonText={`Yes, remove this ${itemName}`}
```
No longer need the `|| 'item'` logic because we are reusing the `itemName` variable that already has that in its instantiation

rhasselle-oddball commented:
can you add screenshot before/after? can you also test this on a form other than simple forms?

gmrabian committed:
once agagin remove for simplicity - logic handled elsewhere

gmrabian commented on src/platform/forms-system/src/js/fields/ArrayField.jsx:
```diff
@@ -389,10 +391,8 @@ export default class ArrayField extends React.Component {
-                          New {uiOptions.itemName || 'item'}
-                        </h3>
+                      {isLast && multipleRows ? (
+                        <h3 className="vads-u-font-size--h5">New {itemName}</h3>
```
No longer need the `|| 'item'` logic because we are reusing the `itemName` variable that already has that in its instantiation

ndsprinkle replied:
I saw this when I was working in here too and got pretty confused. If you follow the trail of commits, it used to use that `itemName` variable, but was later removed. I think that's because `itemName` could theoretically be that aria label function thing which is different than the supplied `uiOptions.itemName`. I think someone got confused in the past and property/variable names got switched up or something. I really don't know. 

Here's Robin adding the `itemName` variable: https://github.com/department-of-veterans-affairs/vets-website/commit/1538e8b16395135c9e97ca7949b9616ed650d7b6#diff-77e9e96767e66321e66766ec6d502a8fa9c8f3f83a3e6fb2e10c9b7bcad66d66R292

And here's obliviga going back to `uiOptions.itemName`, but not adding back the `|| item` piece: https://github.com/department-of-veterans-affairs/vets-website/commit/316e37f9ff585e73325de39aa9d3e7ea00368d7e#diff-77e9e96767e66321e66766ec6d502a8fa9c8f3f83a3e6fb2e10c9b7bcad66d66R287

ndsprinkle replied:
And then I obviously added the `|| item` part back: https://github.com/department-of-veterans-affairs/vets-website/commit/766227d8e2899c40663890674ae196d98b447f1d#diff-77e9e96767e66321e66766ec6d502a8fa9c8f3f83a3e6fb2e10c9b7bcad66d66R394

ndsprinkle replied:
In conclusion, this file is a mess and there have been so many `uiOptions` added over the years that I really have no idea which ones are used where.

And to make things more fun, you've got this file: https://github.com/department-of-veterans-affairs/vets-website/blob/main/src/platform/forms-system/src/js/review/ArrayField.jsx

Which is supposed to the array field that gets rendered on the review page, but it gets **completely bypassed** if you use `keepInPageOnReview`, which 4142 and almost every other form has used to make the array work on the review page.

gmrabian commented on src/platform/forms-system/src/js/fields/ArrayField.jsx:
```diff
@@ -419,19 +419,19 @@ export default class ArrayField extends React.Component {
                           </div>
                           <div className="small-6 right columns">
-                            {index !== 0 && (
+                            {multipleRows && (
```
Allows you to remove the first indexed item if there is more than one item in the array

gmrabian commented:
> can you add screenshot before/after? can you also test this on a form other than simple forms?

@rhasselle-oddball I added screenshots to the description and added a comment about testing another form

gmrabian committed:
Merge branch 'main' into simple-forms/update-array-field

gmrabian commented:
> Unit test failure is happening in main too and is unrelated to these changes

They made a [change to skip it](https://github.com/department-of-veterans-affairs/vets-website/pull/24585)
I've pulled that in so the checks should pass now

gmrabian committed:
refactor to distinguish ui item name and aria item name

gmrabian commented on src/platform/forms-system/src/js/fields/ArrayField.jsx:
```diff
@@ -324,6 +324,7 @@ export default class ArrayField extends React.Component {
       : null;
     const isReviewMode = uiOptions.reviewMode;
     const hasTitleOrDescription = (!!title && !hideTitle) || !!description;
+    const uiItemName = (uiOptions.itemName || 'item').toLowerCase();
```
made reusable `uiItemName` to distinguish between this and the aria item name (which is per array item whereas this one applies to all of them)

gmrabian commented on src/platform/forms-system/src/js/fields/ArrayField.jsx:
```diff
@@ -370,12 +371,11 @@ export default class ArrayField extends React.Component {
             const isRemoving = this.state.removing[index];
             const ariaLabel = uiOptions.itemAriaLabel;
-            const itemName =
+            const ariaItemName =
```
renamed this for clarity

gmrabian committed:
remove debugging code

gmrabian committed:
Merge branch 'main' into simple-forms/update-array-field

gmrabian committed:
Merge branch 'main' into simple-forms/update-array-field

gmrabian committed:
Merge branch 'main' into simple-forms/update-array-field
